Normative: Promise.{all,allSettled,race} should check resolve is a function before opening their iterable

kmiller68 opened the pull request:
This change makes promise.{all,allSettled,race} behave more like a canonical JS implementation.  In particular, the resolve property is now checked to be callable before beginning iteration on the first argument.

Closes #1902.

bakkot commented on spec.html:
```diff
@@ -40003,16 +40007,25 @@ <h1>Promise.all ( _iterable_ )</h1>
+          <emu-alg>
+            1. Let _promiseResolve_ be ? Get(_C_, *"resolve"*).
+            1. If IsCallable(_promiseResolve_) is *false*, throw a *TypeError* exception.
+            1. return _promiseResolve_.
```
```suggestion
            1. Return _promiseResolve_.
```

ljharb commented on spec.html:
```diff
@@ -40003,16 +40007,25 @@ <h1>Promise.all ( _iterable_ )</h1>
+          <h1>Runtime Semantics: GetPromiseResolve ( _promiseCapability_ )</h1>
+          <p>When the GetPromiseResolve abstract operation is call with argument _promiseCapability_, the following steps are taken:</p>
+          <emu-alg>
+            1. Let _promiseResolve_ be ? Get(_C_, *"resolve"*).
```
```suggestion
            1. Assert: IsConstructor(_C_) is *true*.
            1. Let _promiseResolve_ be ? Get(_C_, *"resolve"*).
```

kmiller68 replied:
Done

kmiller68 replied:
Added.

ljharb commented on spec.html:
```diff
@@ -40003,16 +40007,25 @@ <h1>Promise.all ( _iterable_ )</h1>
 
+        <emu-clause id="sec-getpromiseresolve" aoid="GetPromiseResolve">
+          <h1>Runtime Semantics: GetPromiseResolve ( _promiseCapability_ )</h1>
+          <p>When the GetPromiseResolve abstract operation is call with argument _promiseCapability_, the following steps are taken:</p>
```
this should be called with `C` as well.

ljharb commented on spec.html:
```diff
@@ -39990,9 +39990,13 @@ <h1>Promise.all ( _iterable_ )</h1>
+          1. Let _promiseResolve_ be GetPromiseResolve(_promiseCapability_).
+          1. If _promiseResolve_ is an abrupt completion, then
+            1. IfAbruptRejectPromise(_promiseResolve_, _promiseCapability_).
+            1. Return Completion(_promiseResolve_).
```
```suggestion
            1. Return _promiseResolve_.
```
it's already an abrupt completion.

bakkot replied:
This step actually don't make sense - [IfAbruptRejectPromise](https://tc39.es/ecma262/#sec-ifabruptrejectpromise) is a macro, and it causes the invoking function to return if passed an abrupt completion. So control can never reach here.

I think this step can be omitted entirely.

ljharb replied:
That makes much more sense.

kmiller68 replied:
Ah, I didn't realize it was macro! That does indeed make so much more sense...

ljharb commented on spec.html:
```diff
@@ -40073,9 +40086,13 @@ <h1>Promise.allSettled ( _iterable_ )</h1>
+          1. Let _promiseResolve_ be GetPromiseResolve(_promiseCapability_).
+          1. If _promiseResolve_ is an abrupt completion, then
+            1. IfAbruptRejectPromise(_promiseResolve_, _promiseCapability_).
+            1. Return Completion(_promiseResolve_).
```
```suggestion
            1. Return _promiseResolve_.
```
doesn't this mean the promise is rejected, but the function also throws? shouldn't it not throw in this case?

kmiller68 replied:
It worked because IfAbruptRejectPromise is a macro rather than an operation. It's not revenant anymore because I changed the entire if-branch to be a single IfAbruptRejectPromise.

ljharb commented on spec.html:
```diff
@@ -40198,9 +40214,13 @@ <h1>Promise.race ( _iterable_ )</h1>
+          1. Let _promiseResolve_ be GetPromiseResolve(_promiseCapability_).
+          1. If _promiseResolve_ is an abrupt completion, then
+            1. IfAbruptRejectPromise(_promiseResolve_, _promiseCapability_).
+            1. Return Completion(_promiseResolve_).
```
```suggestion
            1. Return _promiseResolve_.
```

kmiller68 commented:
Looks, like @shvaikalesh graciously updated test262 for me with https://github.com/tc39/test262/pull/2648. (Thanks!)

I rebased against the current spec and fixed my summaries to match the newer style. I noticed that we use the phrasing `The abstract operation <Op> takes argument <Arg>.`, which seems odd to me but I kept it for consistency.

ljharb commented on spec.html:
```diff
@@ -40313,9 +40326,11 @@ <h1>Promise.race ( _iterable_ )</h1>
           1. Let _iteratorRecord_ be GetIterator(_iterable_).
           1. IfAbruptRejectPromise(_iteratorRecord_, _promiseCapability_).
-          1. Let _result_ be PerformPromiseRace(_iteratorRecord_, _C_, _promiseCapability_).
+          1. Let _result_ be PerformPromiseRace(_iteratorRecord_, _C_, _promiseCapability_, _promiseResolve_).
```
PerformPromiseRace still only takes 3 arguments

kmiller68 replied:
Ah... sorry. I think that was a bad rebase fixed.

bakkot commented on spec.html:
```diff
@@ -40118,16 +40120,26 @@ <h1>Promise.all ( _iterable_ )</h1>
-          <h1>Runtime Semantics: PerformPromiseAll ( _iteratorRecord_, _constructor_, _resultCapability_ )</h1>
-          <p>The abstract operation PerformPromiseAll takes arguments _iteratorRecord_, _constructor_, and _resultCapability_. It performs the following steps when called:</p>
+          <h1>Runtime Semantics: PerformPromiseAll ( _iteratorRecord_, _constructor_, _resultCapability_, _promiseResolve_ )</h1>
+          <p>The abstract operation PerformPromiseAll takes arguments _iteratorRecord_, _constructor_, _resultCapability_, and _promiseResolve. It performs the following steps when called:</p>
```
```suggestion
          <p>The abstract operation PerformPromiseAll takes arguments _iteratorRecord_, _constructor_, _resultCapability_, and _promiseResolve_. It performs the following steps when called:</p>
```

(The linter doesn't yet know how to catch these, sorry. Someday!)
